Add RFC7845 downmix algorithm

gnattu opened the pull request:
This algorithm is defined in RFC7845 Section 5.1.1.5, supports 2.1, 4.0, 5.0, 5.1, 6.1 and 7.1 channel layouts

Refer: https://datatracker.ietf.org/doc/html/rfc7845#section-5.1.1.5

<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our documentation.
-->

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->

crobibero commented:
Is there a friendlier name for the downmix algorithm?

EraYaN commented:
> Is there a friendlier name for the downmix algorithm?

It's from the Ogg/Opus RFC, but I doubt naming either of those two will be more helpful.

gnattu commented:
There is no "user friendly name" I'm afraid as the RFC document is not a "user friendly" document anyway. I think just use the name RFC7845 is fine as we are using similar name like bt2390 for tonemap algorithms.

crobibero commented on MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:
```diff
@@ -2653,26 +2653,121 @@ public string GetAudioFilterParam(EncodingJobInfo state, EncodingOptions encodin
+                        switch (encodingOptions.DownMixStereoAlgorithm)
+                        {
+                            case DownMixStereoAlgorithms.Rfc7845:
+                                filters.Add("pan=stereo|FL=0.414214*FC+0.585786*FL|FR =0.414214*FC+0.585786*FR");
```
Don't know if it matters but there is extra whitespace in the filter

gnattu replied:
Should not affect functionality but removed for consistency

cvium commented on MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:
```diff
@@ -2653,26 +2653,121 @@ public string GetAudioFilterParam(EncodingJobInfo state, EncodingOptions encodin
                 {
-                    case DownMixStereoAlgorithms.Dave750:
-                        filters.Add("pan=stereo|c0=0.5*c2+0.707*c0+0.707*c4+0.5*c3|c1=0.5*c2+0.707*c1+0.707*c5+0.5*c3");
+                    case 3:
```
Just thinking out loud here, but maybe this would be preferable to nested switches. Thoughts?
```suggestion
                    case 3 when encodingOptions.DownMixStereoAlgorithm == ownMixStereoAlgorithms.Rfc7845:
```

gnattu replied:
I plan to add more algorithms later and will modify this in that one.

cvium commented on MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:
```diff
@@ -2653,26 +2653,121 @@ public string GetAudioFilterParam(EncodingJobInfo state, EncodingOptions encodin
+                        break;
+                    }
+
+                    case 6:
+                    {
+                        switch (encodingOptions.DownMixStereoAlgorithm)
+                        {
+                            case DownMixStereoAlgorithms.Dave750:
+                                filters.Add("pan=stereo|c0=0.5*c2+0.707*c0+0.707*c4+0.5*c3|c1=0.5*c2+0.707*c1+0.707*c5+0.5*c3");
+                                break;
+                            case DownMixStereoAlgorithms.NightmodeDialogue:
+                                filters.Add("pan=stereo|c0=c2+0.30*c0+0.30*c4|c1=c2+0.30*c1+0.30*c5");
+                                break;
+                            case DownMixStereoAlgorithms.Rfc7845:
+                                filters.Add("pan=stereo|c0=0.374107*c2+0.529067*c0+0.458186*c4+0.264534*c5+0.374107*c3|c1=0.374107*c2+0.529067*c1+0.458186*c5+0.264534*c4+0.374107*c3");
+                                break;
+                            case DownMixStereoAlgorithms.None:
+                            default:
+                                break;
+                        }
+
+                        break;
+                    }
```
A longer example here. It reduces the nesting but I can't tell if readability suffers slightly.
```suggestion
                    case 6 when encodingOptions.DownMixStereoAlgorithm == DownMixStereoAlgorithms.Dave750:
                        filters.Add("pan=stereo|c0=0.5*c2+0.707*c0+0.707*c4+0.5*c3|c1=0.5*c2+0.707*c1+0.707*c5+0.5*c3");
                        break;
                    case 6 when encodingOptions.DownMixStereoAlgorithm == DownMixStereoAlgorithms.NightmodeDialogue:
                        filters.Add("pan=stereo|c0=c2+0.30*c0+0.30*c4|c1=c2+0.30*c1+0.30*c5");
                        break;
                    case 6 when encodingOptions.DownMixStereoAlgorithm == DownMixStereoAlgorithms.Rfc7845:
                        filters.Add("pan=stereo|c0=0.374107*c2+0.529067*c0+0.458186*c4+0.264534*c5+0.374107*c3|c1=0.374107*c2+0.529067*c1+0.458186*c5+0.264534*c4+0.374107*c3");
                        break;
                    default:
                        break;
```

gnattu commented:
I refactored the downmix filter related code and split the downmix filter related helpers into a dedicated static class. This makes the algorithms list more extensible and allows for more precise checks to determine if the current settings should fall back to ffmpeg's internal downmix system.

crobibero commented on MediaBrowser.Controller/MediaEncoding/DownMixAlgorithmsHelper.cs:
```diff
@@ -0,0 +1,61 @@
+            5 => "5.0",
+            6 => "5.1", // Could also be 6.0 or hexagonal, prefer 5.1
+            7 => "6.1", // Could also be 7.0, prefer 6.1
+            8 => "7.1", // Could also be 8.0, prefer 6.1
```
```suggestion
            8 => "7.1", // Could also be 8.0, prefer 7.1
```

crobibero commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1733,7 +1733,7 @@ private string GetAudioArguments(StreamState state)
         var channels = state.OutputAudioChannels;
 
-        var useDownMixAlgorithm = state.AudioStream.Channels is 6 && _encodingOptions.DownMixStereoAlgorithm != DownMixStereoAlgorithms.None;
+        var useDownMixAlgorithm = DownMixAlgorithmsHelper.AlgorithmFilterStrings.TryGetValue((_encodingOptions.DownMixStereoAlgorithm, DownMixAlgorithmsHelper.InferChannelLayout(state.AudioStream)), out _);
```
```suggestion
        var useDownMixAlgorithm = DownMixAlgorithmsHelper.AlgorithmFilterStrings.ContainsKey((_encodingOptions.DownMixStereoAlgorithm, DownMixAlgorithmsHelper.InferChannelLayout(state.AudioStream)));
```

crobibero commented on MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:
```diff
@@ -7008,7 +6997,10 @@ public string GetProgressiveVideoAudioArguments(EncodingJobInfo state, EncodingO
 
-            if (channels.HasValue && ((channels.Value != 2 && state.AudioStream?.Channels != 6) || encodingOptions.DownMixStereoAlgorithm == DownMixStereoAlgorithms.None))
+            var useDownMixAlgorithm = state.AudioStream is not null
+                                      && DownMixAlgorithmsHelper.AlgorithmFilterStrings.TryGetValue((encodingOptions.DownMixStereoAlgorithm, DownMixAlgorithmsHelper.InferChannelLayout(state.AudioStream)), out _);
```
```suggestion
                                      && DownMixAlgorithmsHelper.AlgorithmFilterStrings.ContainsKey((encodingOptions.DownMixStereoAlgorithm, DownMixAlgorithmsHelper.InferChannelLayout(state.AudioStream)));
```

Bond-009 commented on MediaBrowser.Controller/MediaEncoding/DownMixAlgorithmsHelper.cs:
```diff
@@ -0,0 +1,61 @@
+            2 => "stereo",
+            3 => "2.1", // Could also be 3.0, prefer 2.1
+            4 => "quad", // Could also be 4.0 (with front center and rear center) and 3.1 with LFE. prefer quad with back left and back right
+            5 => "5.0",
```
Why not prefer 4.1?

gnattu replied:
Because 5.0 is more common? But this is purely subjective and such guessing could be wrong anyway.

Bond-009 replied:
I've personally never seen 5.0, but have seen 4.1
https://www.blu-ray.com/movies/Alien-4K-Blu-ray/230308/

gnattu replied:
We just fixed an issued related to 5.0 in 10.9: https://github.com/jellyfin/jellyfin/issues/11397

gnattu replied:
Search with [4.1 channel resulted in 37 matches](https://www.blu-ray.com/movies/search.php?keyword=&studioid=&videocodec=&disc=&yearfrom=&yearto=&regioncoding=&aspectratio=&aspectratio_original=&releaseyear=&synopsis=&retailerexclusive=&mpaa=&runtimemin=&runtimemax=&videoresolutionid=&sourceid=&subtitles=&audio=4.1&upc=&ean=&asin=&casingid=&slipcoverfront=&slipcoverback=&submit=Search&action=search) and [5.0 resulted in 303 matches](https://www.blu-ray.com/movies/search.php?keyword=&studioid=&videocodec=&disc=&yearfrom=&yearto=&regioncoding=&aspectratio=&aspectratio_original=&releaseyear=&synopsis=&retailerexclusive=&mpaa=&runtimemin=&runtimemax=&videoresolutionid=&sourceid=&subtitles=&audio=5.0&upc=&ean=&asin=&casingid=&slipcoverfront=&slipcoverback=&submit=Search&action=search). I think we should stay with 5.0.

Bond-009 commented on MediaBrowser.Controller/MediaEncoding/DownMixAlgorithmsHelper.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <returns>Channel Layout string.</returns>
+    public static string InferChannelLayout(MediaStream audioStream)
+    {
+        if (!string.IsNullOrWhiteSpace(audioStream.ChannelLayout))
```
`ChannelLayout` should be normalized before being used like this.
For BDMVs it will always be "$CHANNELS.$LFE" and never "stereo" or "quad"

gnattu replied:
> For BDMVs it will always be "$CHANNELS.$LFE" and never "stereo" or "quad"

ffmpeg never outputs 2.0 as "2.0", it will always output stereo

gnattu replied:
And "quad" and "4.0" are actually two different channel layouts. 4.0 means front left, front right ,front center and back center, but quad means front left, front right, back left and back right. All names have their own layout so it is not a normalization problem, it is a correctness problem.

gnattu replied:
If our BDMV parsing code has such normalization, we should revert them. I hope we don't.

Bond-009 replied:
> ffmpeg never outputs 2.0 as "2.0", it will always output stereo

That's bc we don't get that info from ffmpeg for BDMVs, we get it from the playlist info

gnattu replied:
Just looked at the BDMV parser, now we have the problem of unable to identify the correct layout due to lack of information. I personally *want* to cowardly reject such ambiguous layouts and let ffmpeg's internal downmixing system to handle it as ffmpeg knows the layout better than us. The good news is that, we don't need to worry about the stereo case and the quad case is relatively rare compared to more common 5.1 or 7.1.

gnattu replied:
I think we are safe for this case after investigation the current BD implementations. Although dolby digital technically supports 2/2 quad configuration but it is extremely rare and that layout is usually seen on legacy DVDs and music albums. So the 4.0 for BDs would almost always be the 4.0 in ffmpeg definition. I think we can just add a comment to the code just in case there is really an exception.

gnattu replied:
The RFC7845 explicitly requires the quad (2 front 2 rear) layout to use that downmix algorithm, so for 4 channel BDMVs, we will always use ffmpeg's internal system which will be safe enough.

Bond-009 reviewed (approved):
It isn't perfect, but it should be an improvement :+1:
